README
basmat - Bashicu Matrix Calculator

This program calculates Bashicu Matrix. See http://gyafun.jp/ln/basmat.cgi for detail.

(1) Installation

Follow the normal UNIX installation procedure, such as:

     ./configure && make
     su
     make install

(2) Uage

Usage is shown by running

     basmat -h
     
(3) Web interface

Web interface is available at
http://gyafun.jp/ln/basmat.cgi

(4) Author

Fish http://googology.wikia.com/wiki/User:Kyodaisuu


ChangeLog
Bashicu Matrix Calculator Version History

See version history of Bashicu Matrix at
http://googology.wikia.com/wiki/User:Kyodaisuu/BasmatVersion

Version 3.2 (2018-08-24)

    * Added BM 3.2 by Nish (PR#3 by koteitan)
    * Source code of online version is added in cgi directory
    * Legend of ordinal: Buchholz psi to Bashicu's psi

Version 3.1 (2018-08-03)

    * Added BM 2.3 by koteitan (PR#1 by koteitan)
    * Added BM 3.1 by Nish (PR#2 by koteitan)
    * Added -d option for ordinal conversion.
    * Added -T option for test.
    * Default version changed to BM2.
    * Fixed bug of BM version 2.2.

Version 3.0 (2018-06-29)

    * Added support for Bashicu Matrix version 3.

Version 2.1 (2018-03-16)

    * BM version 2.1 and  2.2 implemented by koteitan

Version 2.0 (2017-04-14)

    * Added support for Bashicu Matrix verion 2.
    * Added -d option.
    * Added -h option.
    * Options are handled with getopt.

Version 1.0 (2014-09-01)

    * First release.


test.sh
#!/bin/sh
# Test script to show the differences of BM versions

YABASIC='yabasic'
BASMAT='../basmat' # basmat program should be compiled
BASMAT3='BMsim-20180612.yab' # Tentative version number
SEQUENCE='test.txt' # test sequences

cat $SEQUENCE | while read seq
do
  arg="-v 1 -o 1 -s 30 \"$seq\""
  $BASMAT $arg | head -n 3 > v1.txt
  arg="-v 2 -o 1 -s 30 \"$seq\""
  $BASMAT $arg | head -n 3 > v2.txt
  echo \"$seq\" | $YABASIC $BASMAT3 | head -n 3 > v3.txt

#  DIFF=$(diff v1.txt v2.txt)$(diff v1.txt v3.txt)$(diff v2.txt v3.txt)
  DIFF=$(diff v2.txt v3.txt) # See only the difference of BM2 and BM3
  if [ "$DIFF" != "" ]
  then
    echo "BM1"
    cat v1.txt
    echo "BM2"
    cat v2.txt
    echo "BM3"
    cat v3.txt
  fi
  rm v1.txt v2.txt v3.txt
done


basic/test.sh
#!/bin/sh
# Test script to show the differences of BM versions

YABASIC='yabasic'
BASMAT='../basmat' # basmat program should be compiled
BASMAT2BAS='BM2sim.yab'
BASMAT3BAS='BM3sim.yab'
SEQUENCE='test.txt' # test sequences

cat $SEQUENCE | while read seq
do
  arg="-v 1 -o 1 -s 30 \"$seq\""
  $BASMAT $arg | head -n 3 > v1.txt
  arg="-v 2 -o 1 -s 30 \"$seq\""
  $BASMAT $arg | grep -v "Not" | grep -v "seq" | head -n 3 > v2.txt
  arg="-v 3 -o 1 -s 30 \"$seq\""
  $BASMAT $arg | grep -v "Not" | grep -v "seq" | head -n 3 > v3.txt
  echo \"$seq\" | $YABASIC $BASMAT2BAS | head -n 3 > v2b.txt
  echo \"$seq\" | $YABASIC $BASMAT3BAS | head -n 3 > v3b.txt

#  DIFF=$(diff v1.txt v2.txt)$(diff v1.txt v3.txt)$(diff v2.txt v3.txt)
#  DIFF=$(diff v2.txt v3.txt) # See only the difference of BM2 and BM3
  DIFF=$(diff v2.txt v2b.txt) # See only the difference of BM2 and BM2 basic
#  DIFF=$(diff v3.txt v3b.txt) # See only the difference of BM3 and BM3 basic
  if [ "$DIFF" != "" ]
  then
    # echo "BM1"
    # cat v1.txt
    echo "BM2"
    cat v2.txt
    echo "BM2B"
    cat v2b.txt
    # echo "BM3"
    # cat v3.txt
    # echo "BM3B"
    # cat v3b.txt
  fi
  rm v*.txt
done


basic/BM3.bas
REM Bashicu Matrix version 3
REM Made by Bashicu
REM 21:32, June 12, 2018's version of Bashicu matrix number in this page
REM http://ja.googology.wikia.com/wiki/%E3%83%A6%E3%83%BC%E3%82%B6%E3%83%BC%E3%83%96%E3%83%AD%E3%82%B0:BashicuHyudora/BASIC%E8%A8%80%E8%AA%9E%E3%81%AB%E3%82%88%E3%82%8B%E5%B7%A8%E5%A4%A7%E6%95%B0%E3%81%AE%E3%81%BE%E3%81%A8%E3%82%81

A=9:dim B[∞,∞],C[∞]
for D=0 to 9
 for E=0 to A
  B[1,E]=1
 next
 for F=1 to 0 step -1
  A=A*A
  for G=0 to F
   for H=0 to E
    if B[F-G,H]<B[F,H]-C[H] | B[F,0]=0 then
     if B[F,H+1]=0 then H=E:I=G:G=F else C[H]=B[F,H]-B[F-G,H]
    else
     H=E
    endif
   next
  next
  for J=1 to A
   for K=1 to I
    B[F,0]=B[F-I,0]+C[0]
    for L=1 to E
     B[F,L]=B[F-I,L]
     for M=0 to F
      if B[F-M,0]<B[F,0] | K=1 then
       if B[F-I-M,L]<B[F-M,L] & B[F-M,L]<B[F,L]+C[L] | K=1 then B[F,L]=B[F,L]+C[L]
       M=F
      endif
     next
    next
    F=F+1
   next
  next
  for N=0 to E
   C[N]=0
  next
 next
next
print A


basic/BTMS.bas
REM Bashicu Triangular Matrix
REM Made by Bashicu
REM 20:19, August 20, 2018's version of Bashicu triangular matrix number in this page
REM http://ja.googology.wikia.com/wiki/%E3%83%A6%E3%83%BC%E3%82%B6%E3%83%BC%E3%83%96%E3%83%AD%E3%82%B0:BashicuHyudora/BASIC%E8%A8%80%E8%AA%9E%E3%81%AB%E3%82%88%E3%82%8B%E5%B7%A8%E5%A4%A7%E6%95%B0%E3%81%AE%E3%81%BE%E3%81%A8%E3%82%81

A=99:dim B[∞,∞],C[∞,∞],D[∞]
for E=0 to 99
 for F=1 to A
  B[F,1]=F:C[F,1]=1
  for G=2 to F
   B[F,G]=1
  next
 next
 for H=A to 1 step -1
  A=pow(A,A)
  for I=1 to H-1
   for J=1 to F
    if B[H-I,J]<B[H,J]-D[J] then
     if B[H,J+1]=0 then
      if 1<B[H,J]-B[H-I,J] | (J=2 | K=1) & 0<B[H,J+1] then K=1 else L=I:M=J:I=H:J=F
     else
      if J=1 then D[J]=B[H,J]-B[H-I,J]
     endif
    else
     J=G
    endif
   next
  next
  for N=2 to L
   C[N,1]=1
   for O=2 to F
    for P=0 to H
     if B[H-L+N-P-1,1]<B[H-L+N-1,1] then
      if (C[N-P,O]=1 | 2<O & C[N,O-1]=1) & B[H-L,O]<B[H-L+N-1,O] & O<M then
       C[N,O]=1
      else
       C[N,O]=0
      endif
      P=H
     endif
    next
   next
  next
  for Q=0 to H-1
   if B[H-Q,1]<B[H,1] then
    if M<3 | K=1 then Q=0 elae S=1
    for R=1 to M-1
     D[R]=B[H-Q,R]-B[H-L,R]
    next
    Q=H
   endif
  next
  for T=1 to A
   for U=1+S to L
    for V=1 to F
     if C[U,V]=1 then B[H,V]=B[H-L+S,V]+D[V] else B[H,V]=B[H-L+S,V]
    next
    H=H+1
   next
  next
  K=0:L=0:S=0
  for W=1 to F
   D[W]=0
  next
 next
next
print A


basic/BM2.bas
REM Bashicu Matrix version 2
REM Made by Bashicu
REM 09:17, April 10, 2017's version of Bashicu matrix number in this page
REM http://ja.googology.wikia.com/wiki/%E3%83%A6%E3%83%BC%E3%82%B6%E3%83%BC%E3%83%96%E3%83%AD%E3%82%B0:BashicuHyudora/BASIC%E8%A8%80%E8%AA%9E%E3%81%AB%E3%82%88%E3%82%8B%E5%B7%A8%E5%A4%A7%E6%95%B0%E3%81%AE%E3%81%BE%E3%81%A8%E3%82%81

A=9:dim B[∞,∞],C[∞,∞],D[∞]
for E=0 to 9
 for F=0 to A
  B[1,F]=1:C[1,F]=1
 next
 for G=1 to 0 step -1
  A=A*A
  for H=0 to G
   for I=0 to F
    if B[G-H,I]<B[G,I]-D[I] | B[G,0]=0 then
     if B[G,I+1]=0 then I=F:J=H:H=G else D[I]=B[G,I]-B[G-H,I]
    else
     I=F
    endif
   next
  next
  for K=1 to J
   for L=K to 0 step -1
    if B[G-J+L,0]<B[G-J+K,0] then
     for M=0 to F
      if B[G-J,M]<B[G-J+K,M] & C[L+1,M]=1 then C[K+1,M]=1 else C[K+1,M]=0
     next
     L=0
    endif
   next
  next
  for N=1 to A
   for O=1 to J
    for P=0 to F
     B[G,P]=B[G-J,P]
     if C[O,P]=1 then B[G,P]=B[G,P]+D[P]
    next
    G=G+1
   next
  next
  for Q=1 to F
   D[Q]=0
  next
 next
next
print A
